docs/architecture/adr-004-coroutine-effect-system.md

```diff
@@ -0,0 +1,303 @@
+# ADR 004: Coroutine-Based Effect System for Consensus
+
+## Changelog
+
+* 18-03-2025: Initial version
+
+## Status
+
+Accepted
+
+## Context
+
+The Malachite consensus engine needs to interact with its environment (network, storage, cryptography, application logic) at specific points during execution.
+
+Traditional approaches to handling these interactions include:
+
+1. **Callback-based designs** - Provide callback interfaces that must be implemented by users
+2. **Trait-based polymorphism** - Define traits that users implement to provide required functionality
+3. **Message-passing architectures** - Define protocols for communication between consensus and external components
+
+We needed a design that would:
+- Maintain a clear separation between the consensus algorithm and its environment
+- Keep the consensus code linear and readable despite external interactions
+- Support both synchronous and asynchronous operations
+- Facilitate testing by making effects explicit and mockable
+- Allow different execution environments (sync/async runtimes, actor systems, etc.)
+- Handle errors gracefully without complicating the consensus core
+
+## Decision
+
+We've implemented a **coroutine-based effect system** using the `process!` macro that allows the consensus algorithm to yield control when it needs external resources, and resume when those resources are provided.
+
+### Key Components
+
+1. **`Effect` enum**: A type that represents all possible interactions the consensus algorithm might need from its environment.
+2. **`Resume` enum**: A type that represents all possible ways to resume the consensus algorithm after handling an effect.
+3. **`Resumable` trait**: A trait that connects each effect with its corresponding resume type.
+4. **`process!` macro**: A macro that handles starting the coroutine, processing an input, yielding effects, and resuming consensus with appropriate values.
```

Comment:
Might be useful to define Inputs also as a key component. 

Ref @cason also points out in his review that input definition is missing.